MapboxNavigation/Cache.swift

```diff
@@ -0,0 +1,136 @@
+import Foundation
+
+internal typealias CompletionHandler = () -> Void
+
+/**
+ A cache consists of both in-memory and on-disk components, both of which can be reset.
+ */
+internal protocol BimodalCache {
+    func clearMemory()
+    func clearDisk(completion: CompletionHandler?)
+}
+
+/**
+ A cache which supports storing images
+ */
+internal protocol BimodalImageCache: BimodalCache {
+    func store(_ image: UIImage, forKey key: String, toDisk: Bool, completion completionBlock: CompletionHandler?)
+    func imageFromCache(forKey: String?) -> UIImage?
+}
+
+/**
+ A cache which supports storing data
+ */
+internal protocol BimodalDataCache: BimodalCache {
+    func store(_ data: Data, forKey key: String, toDisk: Bool, completion completionBlock: CompletionHandler?)
+    func dataFromCache(forKey: String?) -> Data?
+}
+
+ /**
+  A general purpose on-disk cache used by both the ImageCache and DataCache implementations
+ */
+internal class FileCache {
+    let diskCacheURL: URL = {
+        let fileManager = FileManager.default
+        let basePath = fileManager.urls(for: .cachesDirectory, in: .userDomainMask).first!
+        let identifier = Bundle.mapboxNavigation.bundleIdentifier!
+        return basePath.appendingPathComponent(identifier + ".downloadedFiles")
+    }()
+
+    let diskAccessQueue = DispatchQueue(label: Bundle.mapboxNavigation.bundleIdentifier! + ".diskAccess")
+    var fileManager: FileManager?
+
+    init() {
+        diskAccessQueue.sync {
+            fileManager = FileManager()
+        }
+    }
+
+    public func store(_ data: Data, forKey key: String, completion: CompletionHandler?) {
+        let dispatchCompletion = {
+            if let completion = completion {
+                DispatchQueue.main.async {
+                    completion()
+                }
+            }
+        }
+        guard let fileManager = fileManager else {
+            dispatchCompletion()
+            return
+        }
+        let cacheURL = diskCacheURL
+        diskAccessQueue.async {
+            self.createCacheDirIfNeeded(cacheURL, fileManager: fileManager)
+            let cacheURL = self.cacheURLWithKey(key)
+
+            do {
+                try data.write(to: cacheURL)
```

Comment:
I would prefer to use our local FileManager for this (since it's already bound to this queue) but had trouble getting it to work as expected and ran out of patience.